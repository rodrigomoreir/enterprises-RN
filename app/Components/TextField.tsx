import React, { 
  useEffect, 
  useRef, 
  useImperativeHandle, 
  forwardRef, 
  useState, 
  useCallback 
} from 'react';
import { 
  View, 
  TextInput, 
  Image, 
  TextInputProps, 
  ViewStyle, 
  StyleProp, 
  TouchableOpacity 
} from 'react-native';
import { getStyles } from './TextField.style'
import { useField } from '@unform/core'
import images from '../Themes/Images';

interface Props extends TextInputProps {
  image?: Object
  placeholder?: string
  name: string
  style?: StyleProp<ViewStyle>
  buttonSearch?: () => void
}

interface InputValueProps {
  value: string
}

interface InputRef {
  focus(): void
}

const TextField: React.RefForwardingComponent<InputRef,Props> = ({ style, image, placeholder, name, buttonSearch, ...rest }, ref ) => {
  const styles = getStyles()
  const inputElementRef = useRef<any>(null)
  const { registerField, defaultValue = '', fieldName, error } = useField(name)
  const inputValueRef = useRef<InputValueProps>({ value: defaultValue })

  const [isFocused, setIsFocused] = useState(false)
  const [isFilled, setIsFilled] = useState(false)

  const handleInputFocus = useCallback(() => {
    setIsFocused(true)
  }, [])

  const handleInputBlur = useCallback(() => {
    setIsFilled(!!inputValueRef.current.value)
  }, [])

  useImperativeHandle(ref, () => ({
    focus() {
      inputElementRef.current.focus()
    }
  }))

  useEffect(() => {
    registerField<string>({
      name: fieldName,
      ref: inputValueRef.current,
      path: 'value',
      setValue(ref: any, value) {
        inputValueRef.current.value = value
        inputElementRef.current.setNativeProps({ text: value })
      },
      clearValue() {
        inputValueRef.current.value = ''
        inputElementRef.current.clear()
      }
    })
  }, [fieldName, registerField])

  return (
    <View style={[
      styles.container, 
      style, 
      isFocused && {borderColor: '#87CEFA'}, 
      !!error && {borderColor: '#C53030'}
    ]}>
      {buttonSearch ? (
        <TouchableOpacity style={styles.buttonSearch} onPress={buttonSearch}>
          <Image source={images.search['24px']} style={styles.imageButton} />
        </TouchableOpacity>
      ) : (
        <Image source={image} style={styles.image} />
      )}
      <TextInput 
        ref={inputElementRef}
        style={styles.textInput} 
        placeholder={placeholder}
        defaultValue={defaultValue}
        onFocus={handleInputFocus}
        onBlur={handleInputBlur}
        onChangeText={value => {
          inputValueRef.current.value = value
        }}
        {...rest}
      ></TextInput>       
    </View>
       
  )
}

export default forwardRef(TextField);
