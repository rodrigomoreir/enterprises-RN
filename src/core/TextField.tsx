import React, {
  useEffect,
  useRef,
  useImperativeHandle,
  forwardRef,
  useState,
  useCallback,
} from 'react'
import { TextInputProps, ViewStyle, StyleProp } from 'react-native'
import styled from 'styled-components/native'
import { screenWidth } from '../utils'
import addOpacity from '../utils/addOpacity'
import { useField } from '@unform/core'
import images from '../assets'
import themeColors from '../theme/theme'

interface Props extends TextInputProps {
  image?: Object
  placeholder?: string
  name: string
  style?: StyleProp<ViewStyle>
  buttonSearch?: () => void
}

interface InputValueProps {
  value: string
}

interface InputRef {
  focus(): void
}

const TextField: React.RefForwardingComponent<InputRef, Props> = (
  { style, image, placeholder, name, buttonSearch, ...rest },
  ref,
) => {
  const inputElementRef = useRef<any>(null)
  const { registerField, defaultValue = '', fieldName, error } = useField(name)
  const inputValueRef = useRef<InputValueProps>({ value: defaultValue })

  const [isFocused, setIsFocused] = useState(false)
  const [isFilled, setIsFilled] = useState(false)

  const handleInputFocus = useCallback(() => {
    setIsFocused(true)
  }, [])

  const handleInputBlur = useCallback(() => {
    setIsFilled(!!inputValueRef.current.value)
  }, [])

  useImperativeHandle(ref, () => ({
    focus() {
      inputElementRef.current.focus()
    },
  }))

  useEffect(() => {
    registerField<string>({
      name: fieldName,
      ref: inputValueRef.current,
      path: 'value',
      setValue(ref: any, value) {
        inputValueRef.current.value = value
        inputElementRef.current.setNativeProps({ text: value })
      },
      clearValue() {
        inputValueRef.current.value = ''
        inputElementRef.current.clear()
      },
    })
  }, [fieldName, registerField])

  return (
    <StyledContainer
      style={[
        style,
        isFocused && { borderColor: themeColors.colors.blue },
        !!error && { borderColor: themeColors.colors.red },
        ,
      ]}>
      {buttonSearch ? (
        <StyledViewSearch onPress={buttonSearch}>
          <StyledImageButton source={images.search['24px']} />
        </StyledViewSearch>
      ) : (
        <StyledImage source={image} />
      )}
      <StyledTextInput
        ref={inputElementRef}
        placeholder={placeholder}
        defaultValue={defaultValue}
        onFocus={handleInputFocus}
        onBlur={handleInputBlur}
        onChangeText={value => {
          inputValueRef.current.value = value
        }}
        {...rest}
      />
    </StyledContainer>
  )
}

const StyledContainer = styled.View`
  width: ${screenWidth - 80}px;
  height: 50px;
  border-color: ${({ theme }) => theme.colors.darkGrey};
  border-width: 1px;
  border-radius: 10px;
  padding-horizontal: 5px;
  flex-direction: row;
  align-items: center;
  margin-vertical: 10px;
`

const StyledViewSearch = styled.TouchableOpacity`
  background-color: ${addOpacity(themeColors.colors.blue, 0.6)};
  width: 55px;
  height: 39px;
  border-top-left-radius: 20px;
  border-bottom-left-radius: 20px;
  align-items: center;
  justify-content: center;
  left: -6px;
`
const StyledImageButton = styled.Image`
  tint-color: ${addOpacity(themeColors.colors.darkGrey, 0.8)};
`
const StyledImage = styled.Image`
  margin-horizontal: 5px;
`
const StyledTextInput = styled.TextInput`
  flex: 1;
`

export default forwardRef(TextField)
