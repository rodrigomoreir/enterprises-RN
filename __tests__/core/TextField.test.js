import React from 'react'

// import renderer from 'react-test-renderer'
import { Form } from '@unform/mobile'

import {
  renderWithTheme,
  matchSnapshotWithTheme,
} from '../../src/functions/testing'

import TextField from '../../src/core/TextField'

describe('TextField', () => {
  test('matches snapshot', () => {
    matchSnapshotWithTheme(
      <Form onSubmit={() => {}}>
        <TextField name={'test'} placeholder={'test'} />
      </Form>,
    )
  })

  test('renders correctly', () => {
    const { getByPlaceholderText } = renderWithTheme(
      <Form onSubmit={() => {}}>
        <TextField name={'test'} placeholder={'test'} onp />
      </Form>,
    )
    const textField = getByPlaceholderText('test')
    expect(textField).toBeTruthy()
  })

  test('renders correctly with button search', () => {
    const onPressMock = jest.fn()
    renderWithTheme(
      <Form onSubmit={() => {}}>
        <TextField
          name={'test'}
          placeholder={'test'}
          buttonSearch={onPressMock}
        />
      </Form>,
    )
  })
})
