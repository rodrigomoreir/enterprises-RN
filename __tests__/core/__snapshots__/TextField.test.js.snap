// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField matches snapshot 1`] = `
<View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderColor": "#696969",
          "borderRadius": 10,
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 50,
          "marginVertical": 10,
          "paddingHorizontal": 5,
          "width": 670,
        },
        undefined,
        false,
        false,
        ,
      ]
    }
  >
    <Image
      style={
        Array [
          Object {
            "marginHorizontal": 5,
          },
        ]
      }
    />
    <TextInput
      allowFontScaling={true}
      defaultValue=""
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholder="test"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
          },
        ]
      }
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;
